File: app/src/main/java/com/sjn/stamp/MusicService.java
package com.sjn.stamp;

import android.content.Intent;
import android.os.Bundle;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaBrowserCompat.MediaItem;
import android.support.v4.media.MediaBrowserServiceCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import com.sjn.stamp.media.playback.Playback;
import com.sjn.stamp.media.playback.PlaybackManager;
import com.sjn.stamp.media.player.CastPlayer;
import com.sjn.stamp.media.player.Player;
import com.sjn.stamp.media.provider.MusicProvider;
import com.sjn.stamp.media.source.LocalMediaSource;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.utils.MediaIDHelper;
import com.sjn.stamp.utils.MediaRetrieveHelper;

import java.util.ArrayList;
import java.util.List;

import static com.sjn.stamp.utils.NotificationHelper.ACTION_CMD;
import static com.sjn.stamp.utils.NotificationHelper.CMD_KILL;
import static com.sjn.stamp.utils.NotificationHelper.CMD_NAME;
import static com.sjn.stamp.utils.NotificationHelper.CMD_PAUSE;
import static com.sjn.stamp.utils.NotificationHelper.CMD_STOP_CASTING;

public class MusicService extends MediaBrowserServiceCompat implements PlaybackManager.PlaybackServiceCallback, Player.PlayerCallback {

    private static final String TAG = LogHelper.makeLogTag(MusicService.class);
    public static final String CUSTOM_ACTION_RELOAD_MUSIC_PROVIDER = "RELOAD_MUSIC_PROVIDER";

    private MediaNotificationManager mMediaNotificationManager;
    private Player mPlayer;
    private CastPlayer mCastPlayer;
    private MusicProvider mMusicProvider;

    @Override
    public void onCreate() {
        super.onCreate();
        LogHelper.d(TAG, "onCreate");
        mMusicProvider = new MusicProvider(this, new LocalMediaSource(this, new MediaRetrieveHelper.PermissionRequiredCallback() {
            @Override
            public void onPermissionRequired() {

            }
        }));
        mMusicProvider.retrieveMediaAsync(
                new MusicProvider.Callback() {
                    @Override
                    public void onMusicCatalogReady(boolean success) {
                        LogHelper.d(TAG, "MusicProvider.callBack start");
                        mPlayer = new Player(MusicService.this, MusicService.this);
                        setSessionToken(mPlayer.initialize(MusicService.this, mMusicProvider));
                        new Thread(new Runnable() {
                            @Override
                            public void run() {
                                mPlayer.restorePreviousState(mMusicProvider);
                                mCastPlayer = new CastPlayer(MusicService.this, mPlayer.getSessionManageListener());
                                try {
                                    mMediaNotificationManager = new MediaNotificationManager(MusicService.this);
                                } catch (RemoteException e) {
                                    throw new IllegalStateException("Could not create a MediaNotificationManager", e);
                                }
                                LogHelper.d(TAG, "MusicProvider.callBack end");
                            }
                        }).start();
                    }
                }
        );
    }

    @Override
    public int onStartCommand(Intent startIntent, int flags, int startId) {
        LogHelper.d(TAG, "onStartCommand");
        if (startIntent != null) {
            String action = startIntent.getAction();
            String command = startIntent.getStringExtra(CMD_NAME);
            if (ACTION_CMD.equals(action)) {
                handleActionCommand(command);
            } else if (mPlayer != null) {
                mPlayer.handleIntent(startIntent);
            }
        }
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        LogHelper.d(TAG, "onDestroy");
        if (mPlayer != null) {
            mPlayer.stop();
        }
        if (mMediaNotificationManager != null) {
            mMediaNotificationManager.stopNotification();
        }
        if (mCastPlayer != null) {
            mCastPlayer.finish();
        }
    }

    /**
     * {@link MediaBrowserServiceCompat}
     */
    @Override
    public void onCustomAction(@NonNull String action, Bundle extras,
                               @NonNull Result<Bundle> result) {
        LogHelper.d(TAG, "onCustomAction " + action);
        switch (action) {
            case CUSTOM_ACTION_RELOAD_MUSIC_PROVIDER:
                result.detach();
                mMusicProvider.cacheAndNotifyLatestMusicMap();
                return;
        }
        result.sendError(null);
    }

    @Override
    public void onSearch(@NonNull final String query, Bundle extras,
                         final @NonNull Result<List<MediaBrowserCompat.MediaItem>> result) {
        LogHelper.d(TAG, "onSearch " + query);
        result.detach();
        new Thread(new Runnable() {
            @Override
            public void run() {
                result.sendResult(mMusicProvider.getChildren(query, getResources()));
            }
        }).start();
    }

    @Override
    public BrowserRoot onGetRoot(@NonNull String clientPackageName, int clientUid,
                                 Bundle rootHints) {
        LogHelper.d(TAG, "OnGetRoot: clientPackageName=" + clientPackageName,
                "; clientUid=" + clientUid + " ; rootHints=", rootHints);
        return new BrowserRoot(MediaIDHelper.MEDIA_ID_ROOT, null);
    }

    @Override
    public void onLoadChildren(@NonNull final String parentMediaId,
                               @NonNull final Result<List<MediaItem>> result) {
        LogHelper.d(TAG, "OnLoadChildren: parentMediaId=", parentMediaId);
        if (MediaIDHelper.MEDIA_ID_EMPTY_ROOT.equals(parentMediaId)) {
            result.sendResult(new ArrayList<MediaItem>());
        } else if (mMusicProvider.isInitialized()) {
            result.detach();
            new Thread(new Runnable() {
                @Override
                public void run() {
                    result.sendResult(mMusicProvider.getChildren(parentMediaId, getResources()));
                }
            }).start();
        } else {
            result.detach();
        }
    }

    /**
     * {@link PlaybackManager.PlaybackServiceCallback}
     */
    @Override
    public void onPlaybackStart() {
        mPlayer.setActive(true);
        // The service needs to continue running even after the bound client (usually a
        // MediaController) disconnects, otherwise the music playback will stop.
        // Calling startService(Intent) will keep the service running until it is explicitly killed.
        startService(new Intent(getApplicationContext(), MusicService.class));
        if (mMediaNotificationManager != null) {
            mMediaNotificationManager.startForeground();
        }
    }

    @Override
    public void onPlaybackStop() {
        mPlayer.setActive(false);
        if (mMediaNotificationManager != null) {
            mMediaNotificationManager.stopForeground(true);
        }
    }

    @Override
    public void onPlaybackStateUpdated(PlaybackStateCompat newState) {
        mPlayer.setPlaybackState(newState);
    }

    @Override
    public void onNotificationRequired() {
        mMediaNotificationManager.startNotification();
    }

    /**
     * {@link Player.PlayerCallback}
     */
    @Override
    public Playback requestPlayback(Playback.Type type) {
        LogHelper.d(TAG, "requestPlayback");
        return type.createInstance(mMusicProvider, this);
    }

    private void handleActionCommand(String command) {
        switch (command) {
            case CMD_PAUSE:
                mPlayer.pause();
                break;
            case CMD_STOP_CASTING:
                if (mCastPlayer != null) {
                    mCastPlayer.stop();
                }
                break;
            case CMD_KILL:
                if (mPlayer != null) {
                    stopSelf();
                }
                break;
        }
    }

}

File: app/src/main/java/com/sjn/stamp/MusicService.kt
package com.sjn.stamp

import android.app.Service
import android.content.Intent
import android.os.Bundle
import android.os.RemoteException
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaBrowserCompat.MediaItem
import android.support.v4.media.MediaBrowserServiceCompat
import android.support.v4.media.session.MediaControllerCompat
import com.sjn.stamp.media.notification.NotificationManager
import com.sjn.stamp.media.playback.Playback
import com.sjn.stamp.media.playback.PlaybackManager
import com.sjn.stamp.media.provider.MusicProvider
import com.sjn.stamp.media.source.LocalMediaSource
import com.sjn.stamp.ui.observer.MediaControllerObserver
import com.sjn.stamp.utils.*
import java.util.*

class MusicService : MediaBrowserServiceCompat(), PlaybackManager.PlaybackServiceCallback {

    companion object {
        private val TAG = LogHelper.makeLogTag(MusicService::class.java)
        const val NOTIFICATION_CMD_PLAY = "CMD_PLAY"
        const val NOTIFICATION_CMD_PAUSE = "CMD_PAUSE"
        const val NOTIFICATION_CMD_STOP_CASTING = "CMD_STOP_CASTING"
        const val NOTIFICATION_CMD_KILL = "CMD_KILL"
        const val CUSTOM_ACTION_RELOAD_MUSIC_PROVIDER = "RELOAD_MUSIC_PROVIDER"
        const val CUSTOM_ACTION_SET_QUEUE = "SET_QUEUE"
        const val CUSTOM_ACTION_SET_QUEUE_BUNDLE_KEY_TITLE = "SET_QUEUE_BUNDLE_KEY_TITLE"
        const val CUSTOM_ACTION_SET_QUEUE_BUNDLE_MEDIA_ID = "SET_QUEUE_BUNDLE_MEDIA_ID"
        const val CUSTOM_ACTION_SET_QUEUE_BUNDLE_KEY_QUEUE = "SET_QUEUE_BUNDLE_KEY_QUEUE"
    }

    private var playOnPrepared = false

    private var playbackManager: PlaybackManager? = null
    private var mediaController: MediaControllerCompat? = null

    private val notificationManager = NotificationManager(this)
    private val musicProvider = MusicProvider(this,
            LocalMediaSource(this, object : MediaRetrieveHelper.PermissionRequiredCallback {
                override fun onPermissionRequired() {
                }
            }))

    override fun onCreate() {
        super.onCreate()
        LogHelper.d(TAG, "onCreate")
        musicProvider.retrieveMediaAsync(object : MusicProvider.Callback {
            override fun onMusicCatalogReady(success: Boolean) {
                LogHelper.d(TAG, "MusicProvider.callBack start")
                playbackManager = PlaybackManager(this@MusicService, this@MusicService, musicProvider, Playback.Type.LOCAL).apply {
                    restorePreviousState()
                }
                sessionToken = playbackManager?.sessionToken
                sessionToken?.let {
                    mediaController = MediaControllerCompat(this@MusicService, it).apply {
                        MediaControllerObserver.register(this)
                    }
                }
                PlayModeHelper.restore(this@MusicService)
                try {
                    notificationManager.updateSessionToken()
                } catch (e: RemoteException) {
                    throw IllegalStateException("Could not create a NotificationManager", e)
                }
                if (playOnPrepared) {
                    playOnPrepared = false
                    mediaController?.transportControls?.play()
                }
                LogHelper.d(TAG, "MusicProvider.callBack end")
            }
        })
    }

    override fun onDestroy() {
        LogHelper.d(TAG, "onDestroy")
        mediaController?.let {
            it.transportControls.stop()
            MediaControllerObserver.unregister(it)
        }
        notificationManager.stopNotification()
    }

    override fun onStartCommand(startIntent: Intent?, flags: Int, startId: Int): Int {
        LogHelper.d(TAG, "onStartCommand")
        startIntent?.let {
            if (NotificationHelper.ACTION_CMD == it.action) {
                handleActionCommand(it.getStringExtra(NotificationHelper.CMD_NAME))
            } else {
                playbackManager?.handleIntent(it)
            }
        }
        return Service.START_NOT_STICKY
    }

    /**
     * [MediaBrowserServiceCompat]
     */
    override fun onCustomAction(action: String, extras: Bundle?,
                                result: MediaBrowserServiceCompat.Result<Bundle>) {
        LogHelper.d(TAG, "onCustomAction $action")
        when (action) {
            CUSTOM_ACTION_RELOAD_MUSIC_PROVIDER -> {
                result.detach()
                musicProvider.cacheAndNotifyLatestMusicMap()
                return
            }
            CUSTOM_ACTION_SET_QUEUE -> {
                result.detach()
                extras?.let {
                    playbackManager?.startNewQueue(
                            extras.getString(CUSTOM_ACTION_SET_QUEUE_BUNDLE_KEY_TITLE),
                            extras.getString(CUSTOM_ACTION_SET_QUEUE_BUNDLE_MEDIA_ID),
                            extras.getParcelableArrayList(CUSTOM_ACTION_SET_QUEUE_BUNDLE_KEY_QUEUE))
                }
                return
            }
        }
        result.sendError(null)
    }

    override fun onSearch(query: String, extras: Bundle?,
                          result: MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>>) {
        LogHelper.d(TAG, "onSearch $query")
        result.detach()
        if (musicProvider.isInitialized) {
            Thread(Runnable { result.sendResult(musicProvider.getChildren(query, resources)) }).start()
        }
    }

    override fun onGetRoot(clientPackageName: String, clientUid: Int,
                           rootHints: Bundle?): MediaBrowserServiceCompat.BrowserRoot? {
        LogHelper.d(TAG, "OnGetRoot: clientPackageName=$clientPackageName", "; clientUid=$clientUid ; rootHints=", rootHints)
        return MediaBrowserServiceCompat.BrowserRoot(MediaIDHelper.MEDIA_ID_ROOT, null)
    }

    override fun onLoadChildren(parentMediaId: String,
                                result: MediaBrowserServiceCompat.Result<List<MediaItem>>) {
        LogHelper.d(TAG, "OnLoadChildren: parentMediaId=", parentMediaId)
        if (MediaIDHelper.MEDIA_ID_EMPTY_ROOT == parentMediaId) {
            result.sendResult(ArrayList())
        } else {
            result.detach()
            if (musicProvider.isInitialized) {
                Thread(Runnable { result.sendResult(musicProvider.getChildren(parentMediaId, resources)) }).start()
            }
        }
    }

    /**
     * [PlaybackManager.PlaybackServiceCallback]
     */
    override fun onPlaybackStart() {
        LogHelper.d(TAG, "onPlaybackStart")
        // The service needs to continue running even after the bound client (usually a
        // MediaController) disconnects, otherwise the music playback will stop.
        // Calling startService(Intent) will keep the service running until it is explicitly killed.
        startForegroundServiceCompatible(Intent(applicationContext, MusicService::class.java))
        notificationManager.startForeground()
    }

    override fun onPlaybackStop() {
        LogHelper.d(TAG, "onPlaybackStop")
        notificationManager.stopForeground(false)
    }

    override fun onNotificationRequired() {
        LogHelper.d(TAG, "onNotificationRequired")
        notificationManager.startNotification()
    }

    private fun handleActionCommand(command: String) {
        LogHelper.d(TAG, "handleActionCommand ", command)
        when (command) {
            NOTIFICATION_CMD_PLAY -> mediaController?.transportControls?.play()
                    ?: run { playOnPrepared = true }
            NOTIFICATION_CMD_PAUSE -> mediaController?.transportControls?.pause()
            NOTIFICATION_CMD_STOP_CASTING -> playbackManager?.stopCasting()
            NOTIFICATION_CMD_KILL -> playbackManager?.let { stopSelf() }
        }
    }

}
